### DISCUSSION TITLE
SECCDC Quals Feedback (2024)

### DISCUSSION THREAD

**Discussion by KCarretto** [spellshift/realm#568] [Ideas]
A place to record feedback for SECCDC Quals realm usage.

# Missing Eldritch Features
* Ensure we include a list of all starlark provided functionality in eldritch documentation
* File Permissions (chattr, chmod, lsattr)
    * Be able to set permissions during file.write() 
* Regex (e.g. replace with regex)
* globbing (*)

**Comment by KCarretto** [comment]
We should have a defacto "test" group so that prebuilt dashboards can prefilter these out

**Comment by KCarretto** [comment]
File list tome says "Error: Invalid Path" when file doesn't exist, we should update this to say the file doesn't exist

**Comment by hulto** [comment]
Feature request: show output size in quest output to allow fast checks on how much data is back.

**Comment by hulto** [comment]
Pain point: User didn't realize that create quests filters out dead beacons.

**Comment by hulto** [comment]
Feature request: SOCK5 proxy or at least a tome to drop chisel.

**Comment by KCarretto** [comment]
Feature: Personal quest view (e.g. list quests that I've created), likely a "My Quests" sidenav option

**Comment by KCarretto** [comment]
Feature: View of recently killed beacons (answer the question "what have I lost recently")

**Comment by KCarretto** [comment]
Feature: Preserve filters when refreshing page (or add polling), noticed on hosts view

**Comment by KCarretto** [comment]
Feature: Maybe an eval tome? People a little sad with the lack of tomes available

**Comment by KCarretto** [comment]
Feature: Consider enabling multiple tomes to be queued for a quest, it could be useful to maintain context (e.g. I want to do this thing <QuestName> across these hosts).

**Reply by hulto** [reply]
An alternative @cmp5987 and I discussed is adding "run and task more beacons" button on the quest output or at the end of the quest creation flow.
The button would preserver state of selected beacons.

**Comment by KCarretto** [comment]
Feature: Filter quest view by creator

**Comment by KCarretto** [comment]
Feature: Track metrics on success rates by Tome

**Comment by KCarretto** [comment]
Feature: Recon Tome
* Process list / netstat
* User list
* Report common service configs as structured file output to tavern (SSH, Apache, etc.)

**Comment by KCarretto** [comment]
Feature: Consider merging Quest History and Quest Outputs view, if we add enhanced search to Quest History (e.g. searching output) then we wouldn't require the top-level Quest Outputs view

**Comment by KCarretto** [comment]
Feature: Include Process Group ID (pgid) in results of process.list

**Comment by nullmonk** [comment]
Bugfix: The `print` function does not add a newline at the end. This goes against the [starlark spec](https://github.com/google/starlark-go/blob/master/doc/spec.md#print)

**Comment by nullmonk** [comment]
Bugfix/feature: Immutable bit does not show up in `permissions` on file objects returned by `file.list`

**Comment by nullmonk** [comment]
Discussion: `user` and `group` are returning different values in file objects returned by `file.list`. User returns the user as a string, Group is the GID. Should they both be GID/UID or both be a string?

**Reply by nullmonk** [reply]
Followup, group should be a string, as `sys.get_user()` returns the groups as strings. Since there is no way to convert the group name to an int, it is impossible to tell if the user has the ability to read a file via file.permissions

**Comment by hulto** [comment]
Ability to cancel jobs if they haven't been picked up yet

**Comment by hulto** [comment]
An easier way to test a quest before rolling it out.
Maybe if we have a test team provide a run on test team option to the quest creation view.

**Comment by hulto** [comment]
pspy tome https://github.com/DominicBreuker/pspy

**Reply by nullmonk** [reply]
How do you want this to work? Each new proc gets queued to be sent to the c2? Or a run-and-done and dump the output?

**Reply by hulto** [reply]
Two implementations in mind:
1. Re-implement the inotify watcher as an eldritch function. The function would take a function as an input param which would handle reporting the data. A simple example being: `sys.pspy(func_handle=print)` which would just print the string rep of the proc.

2. Write a tome that copies out the real pspy, execute it with disown, use `file.follow` on the output file to grab everything.

**Comment by hulto** [comment]
Desktop streaming / long running tome running screenshot

**Comment by hulto** [comment]
The copy pasting the input params in task details

**Comment by nullmonk** [comment]
Ability to pass input params in golem for testing tomes

**Comment by hulto** [comment]
Docs to backup and restore the SQL database.
Also maybe docs on import the SQL DB into a local dev instance. (And manually approving your user)

**Comment by hulto** [comment]
Code freeze on the 26th leading up to ISTS?

**Comment by hulto** [comment]
oauth state errors on first login

**Comment by KCarretto** [comment]
Naming for file transfers:

* Files reported by the agent to the c2 (currently `report.file()` in eldritch, `report_file` in gRPC)
* Files requested by the agent (`assets.copy()` in eldritch, `download_file` in gRPC)

**Comment by hulto** [comment]
Do reusable helper functions belong in tomes, as eldritch functions, or in a "common" tome library.
